### PR TITLE
Ga time on page

### LINKED ISSUE #3967: GA: find a reliable way to track time on page for single page sessions
In the past we have relied on the triggering of scroll depth events to give us a proxy for the time on page. This was removed, but rather than reinstate what was an inaccurate method we want to implement something better.

Potential solution is to use the beacon API.

### PR DESCRIPTION
**Pull request by gestchild**
Fixes #3967

### CODE REVIEW

File: common/views/pages/_app.js [hunk 1]
```diff
@@ -13,29 +13,43 @@ import ErrorPage from '../../views/components/ErrorPage/ErrorPage';
...
 
-function triggerEngagement() {
-  ReactGA.event({
+function trackVisibleTimeOnPage () {
+  const visibleTime = Math.round(window.performance.now() - previousTimeOnPage - pageHiddenTime);
+  trackEvent({
     category: 'Engagement',
-    action: 'Time on page >=',
-    label: '10 seconds'
+    action: 'Time page is visible',
```
Comment (davidpmccormick) [inline common/views/pages/_app.js]: nitpick, but I think the rest of our new actions don't start with a capital letter?

File: common/utils/ga.js [hunk 2]
```diff
@@ -17,7 +17,10 @@ export function trackIfOutboundLink(url: string) {
 export type GaEvent = {|
   category: string,
   action: string,
-  label: string
+  label?: string,
+  value?: string | number,
```
Comment (davidpmccormick) [inline common/utils/ga.js]: is `value` allowed to be a string for GA? I thought I maybe read somewhere that it _had_ to be a number. Not sure though.

Comment (davidpmccormick) [reply common/utils/ga.js]: https://stackoverflow.com/a/26250714

Comment (gestchild) [reply common/utils/ga.js]: you're right

File: common/views/pages/_app.js [hunk 3]
```diff
@@ -87,6 +107,16 @@ export default class WecoApp extends App {
       titleCase: false
     }]);
 
+    try {
+      if (document.hidden) {
+        pageVisibilityLastChanged = window.performance.now(); // in case page is opened in a new tab
+      }
+      document.addEventListener('visibilitychange', calculateHiddenTimeOnPage, false);
```
Comment (jamesgorrie) [inline common/views/pages/_app.js]: Is there a reason to set the useCapture here to false, where that's the default?

File: common/views/pages/_app.js [hunk 4]
```diff
@@ -13,29 +13,43 @@ import ErrorPage from '../../views/components/ErrorPage/ErrorPage';
...
 let globalAlert;
-let engagement;
+let previousTimeOnPage = 0;
+let pageHiddenTime = 0;
+let pageVisibilityLastChanged = 0;
 
-function triggerEngagement() {
-  ReactGA.event({
+function trackVisibleTimeOnPage () {
+  const visibleTime = Math.round(window.performance.now() - previousTimeOnPage - pageHiddenTime);
```
Comment (jamesgorrie) [inline common/views/pages/_app.js]: it's a tad confusing here using the word `Time`, which could indicate accrued time, or the epoch time.
Might be worth disambiguating?

File: common/utils/ga.js [hunk 5]
```diff
@@ -17,7 +17,10 @@ export function trackIfOutboundLink(url: string) {
 export type GaEvent = {|
   category: string,
   action: string,
-  label: string
+  label?: string,
+  value?: number,
+  nonInteraction?: boolean,
+  transport?: string
```
Comment (jamesgorrie) [inline common/utils/ga.js]: Could this be enumerated?

File: common/views/pages/_app.js [hunk 6]
```diff
@@ -87,6 +107,16 @@ export default class WecoApp extends App {
...
     }]);
 
+    try {
+      if (document.hidden) {
+        pageVisibilityLastChanged = window.performance.now(); // in case page is opened in a new tab
+      }
+      document.addEventListener('visibilitychange', calculateHiddenTimeOnPage, false);
+      window.addEventListener('beforeunload', trackVisibleTimeOnPage);
+    } catch (error) {
+      // nada
```
Comment (jamesgorrie) [inline common/views/pages/_app.js]: Is this for unsupported browsers?

If so, might be worth sending an event to GA, could be interesting to see how many browsers we could have tracked, but didn't.

File: common/utils/ga.js (hunk 5 above)
Comment (jamesgorrie) [inline common/utils/ga.js]:
```suggestion
  transport?: 'beacon'
```

**Pull request merged**